client/app/src/App.js

```diff
@@ -1,14 +1,76 @@
-import Amplify from 'aws-amplify';
 import React from 'react';
 import { Platform, StatusBar, StyleSheet, View } from 'react-native';
+import { Provider } from 'react-redux';
+import { PersistGate } from 'redux-persist/integration/react';
+import Amplify from 'aws-amplify';
+import { withAuthenticator } from 'aws-amplify-react-native';
 import EStyleSheet from 'react-native-extended-stylesheet';
 import { DefaultTheme, Provider as PaperProvider } from 'react-native-paper';
-import { withAuthenticator } from 'aws-amplify-react-native';
-
+import { persistor, store, actions } from './state';
+import concertApi from './api/concertApi';
+import NavigationService from './navigation/NavigationService';
 import AppNavigator from './navigation/AppNavigator';
 import layout from './constants/Layout';
 import config from './config';
 
+// ** Event listeners ** //
+const handleStageNavigation = stageId => {
+  console.log('in handleStageNavigation');
+  const currentState = store.getState();
+  const { choiceType } = currentState;
+
+  let screen = 'Welcome';
+  if (stageId === 'STAGE_CHOICE_IMAGERY') {
+    screen = 'MentalImagery';
+  } else if (stageId === 'STAGE_CHOICE_COLOR_EMOTION') {
+    if (choiceType === 'CHOICE_COLOR') {
+      screen = 'Colors';
+    } else {
+      screen = 'Emotions';
+    }
+  } else if (stageId === 'STAGE_CHOICE_CHILLS') {
+    screen = 'Chills';
+  } else if (stageId === 'STAGE_END') {
+    screen = 'Results';
+  } else {
+    screen = 'Welcome'; // temporary
+    console.log('stub: something went wrong in handleStageNavigation');
+    // something went wrong
+    // navigate to 'something went wrong screen'?
+  }
+  store.dispatch(actions.setLastKnownScreen(screen));
+  NavigationService.navigate(screen);
+};
+
+concertApi.on('WEBSOCKET_CONNECTED', data => {
+  console.log('stream in ws connected is:', data);
+  try {
+    const { stageId, choiceType, choiceInverted } = data;
+    console.log('choiceType in WS connected is:', choiceType);
+    store.dispatch(actions.setChoiceType(choiceType));
+    store.dispatch(actions.setChoiceInverted(choiceInverted));
+    handleStageNavigation(stageId);
```

Comment:
This seems to be discarding all the stage-specific data and only preserving stageId. Do you have a plan for how to persist that additional information?

Reply:
You're right, it currently is. Will the data for one stage need to be stored after the concert progresses to the next stage? 

If it will need to be kept, then the stage data could be stored as a stageAttributes property of the root redux state, with keys that correspond to each stageId. E.g., to access the stage-specific data for STAGE_CHOICE_CHILLS would be accessed by state.stageAttributes.STAGE_CHOICE_CHILLS. And the attributes for the current stage could be accessed by state.stageAttributes[state.stageId] (see my responses to your comment above).

If the data can be discarded after progression to the next stage, then the stage-specific data could be stored by in state.stageAttributes that is overwritten on progression to the next stage. 

What do you think?

Reply:
At the end of the app-guided part of the concert (during the intermission), the user should be presented with the "results screen" which shows their choices compared to the aggregate. The aggregate choices will be provided by the back-end, but the user's choices need to be persisted in the app to be displayed in this section. Where this information is stored, whether it's in a "stage-specific" part of the data store, is up to you. Does the app need to remember that the last stage was frequency 20 and timeout 5? No.